pkg/catalog/catalog.go

```diff
@@ -161,21 +174,49 @@ func newSharedCatalog() (*Catalog, error) {
 	return c, nil
 }
 
-// getCatalogCache retrieves the catalog from the local directory.
-func getCatalogCache() (catalog *Catalog, err error) {
-	b, err := os.ReadFile(getCatalogCachePath())
+// getCatalogCache retrieves the catalog from the local directory along with locking the catalog file
+// If `setWriteLock` is false, it will read the catalog file with ReadLock and release the lock at the same time
+// If `setWriteLock` is true, it will apply WriteLock to the catalog file, read the catalog file
+// and keep the WriteLock to the file along with returning `unlock` function. It is caller
+// functions responsibility to unlock the WriteLock after the catalog update
+func getCatalogCache(setWriteLock bool) (catalog *Catalog, unlock func(), err error) {
+	unlock = func() {}
+	var b []byte
+
+	if setWriteLock {
+		var lockedFile *lockedfile.File
+		if utils.PathExists(getCatalogCachePath()) {
+			lockedFile, err = lockedfile.Edit(getCatalogCachePath())
+		} else {
+			// Create directory path if missing before locking the file
+			_ = os.MkdirAll(filepath.Dir(getCatalogCachePath()), 0777)
+			lockedFile, err = lockedfile.Create(getCatalogCachePath())
+		}
```

Comment:
Please notice this latest change where I am using 2 different functions based on the existence of the catalog file. cc @marckhouzam @vuil

Reply:
Oh, nice find! 
Did the test fail, or did you just notice by looking at the code?  I'm asking because I can't find a way to cause a problem before this change

Reply:
One of the unit tests failed which was testing the catalog update when no file or directory existed.